tools/src/wyvern/tools/ArchitectureInterpreter.java

```diff
@@ -0,0 +1,321 @@
+                    portObjs.add((ASTPortDecl) portDecls.get(pair[1]));
+                }
+                // Load the connector type module and get context
+                Module m = state.getResolver().resolveType(connector + "Properties");
+                GenContext genCtx = Globals.getGenContext(state);
+                EvalContext evalCtx = Globals.getStandardEvalContext();
+                SeqExpr mSeqExpr = state.getResolver().wrapWithCtx(m.getExpression(),
+                        m.getDependencies(), evalCtx);
+                genCtx = mSeqExpr.extendContext(genCtx);
+                evalCtx = mSeqExpr.interpretCtx(evalCtx).getSecond();
+                // Get AST node
+                TypedModuleSpec mSpec = m.getSpec();
+                ValueType mType = mSpec.getType();
+                StructuralType mSType = mType.getStructuralType(genCtx);
+                ConcreteTypeMember connectorDecl = (ConcreteTypeMember) mSType
+                        .findDecl(connector + "Properties", genCtx);
+                // Interpret type and get metadata
+                ConcreteTypeMember contype = (ConcreteTypeMember) connectorDecl
+                        .interpret(evalCtx);
+                Value metadata = contype.getMetadataValue();
+                ValueType metadataType = metadata.getType();
+                StructuralType metadataStructure = metadataType
+                        .getStructuralType(genCtx);
+
+                // Execute metadata
+                Value portCompatibility = null, connectorImpl = null,
+                        connectorInit = null;
+                for (DeclType dt : metadataStructure.getDeclTypes()) {
+                    if (dt instanceof DefDeclType) {
+                        DefDeclType defdecl = (DefDeclType) dt;
+                        String methodName = defdecl.getName();
+                        List<Value> testArgs = new LinkedList<Value>();
+                        if (methodName.equals("checkPortCompatibility")) {
+                            testArgs.add(javaToWyvernList(portObjs));
+                            portCompatibility = ((Invokable) metadata)
+                                    .invoke(methodName, testArgs).executeIfThunk();
+                        } else if (methodName.equals("generateConnectorImpl")) {
+                            testArgs.add(javaToWyvernList(portObjs));
+                            connectorImpl = ((Invokable) metadata)
+                                    .invoke(methodName, testArgs).executeIfThunk();
+                        } else if (methodName.equals("generateConnectorInit")) {
+                            connectorInit = portCompatibility;
+                        } else {
+                            ToolError.reportError(ErrorMessage.INVALID_CONNECTOR_METADATA,
+                                    FileLocation.UNKNOWN, connector);
+                        }
+                    }
+                }
```

Comment:
You have conflated the check for existence of the required functions in the metadata with calling them. A side effect of that is that the generateConnectorImpl() function might get called before the checkPortCompatibility() function if it is declared first. Please don't make assumptions about the order in which the functions are declared and invoke them in the proper order.

Also, why are you assigning portCompatibility to connectorInit in line 158?